invokeai/app/services/download/download_default.py

```diff
@@ -0,0 +1,367 @@
+# Copyright (c) 2023, Lincoln D. Stein
+"""Implementation of multithreaded download queue for invokeai."""
+
+import os
+import re
+import threading
+import traceback
+from logging import Logger
+from pathlib import Path
+from queue import PriorityQueue
+from typing import Dict, List, Optional, Set, Union
+
+import requests
+from pydantic.networks import AnyHttpUrl, Url
+from requests import HTTPError
+from tqdm import tqdm
+
+from invokeai.app.services.events.events_base import EventServiceBase
+from invokeai.app.util.misc import get_iso_timestamp
+from invokeai.backend.util.logging import InvokeAILogger
+
+from .download_base import (
+    DownloadEventHandler,
+    DownloadJob,
+    DownloadJobCancelledException,
+    DownloadJobStatus,
+    DownloadQueueServiceBase,
+    UnknownJobIDException,
+)
+
+# Maximum number of bytes to download during each call to requests.iter_content()
+DOWNLOAD_CHUNK_SIZE = 100000
+
+# marker that the queue is done and that thread should exit
+STOP_JOB = DownloadJob(id=-99, priority=-99, source=Url("http://dummy.com/"), dest=Path("/"))
+
+
+class DownloadQueueService(DownloadQueueServiceBase):
+    """Class for queued download of models."""
+
+    _jobs: Dict[int, DownloadJob]
+    _worker_pool: Set[threading.Thread]
+    _queue: PriorityQueue
+    _lock: threading.Lock
+    _logger: Logger
+    _events: Optional[EventServiceBase] = None
+    _next_job_id: int = 0
+    _requests: requests.sessions.Session
+
+    def __init__(
+        self,
+        max_parallel_dl: int = 5,
+        event_bus: Optional[EventServiceBase] = None,
+        requests_session: Optional[requests.sessions.Session] = None,
+    ):
+        """
+        Initialize DownloadQueue.
+
+        :param max_parallel_dl: Number of simultaneous downloads allowed [5].
+        :param requests_session: Optional requests.sessions.Session object, for unit tests.
+        """
+        self._jobs = {}
+        self._next_job_id = 0
+        self._queue = PriorityQueue()
+        self._worker_pool = set()
+        self._lock = threading.Lock()
+        self._logger = InvokeAILogger.get_logger("DownloadQueueService")
+        self._event_bus = event_bus
+        self._requests = requests_session or requests.Session()
+
+        self._start_workers(max_parallel_dl)
+
+    def __del__(self) -> None:
+        if any(thread.is_alive for thread in self._worker_pool):
+            self._queue.put(STOP_JOB)
```

Comment:
I think you're already aware of this from conversations on the previous PR, but commenting to remind myself:
- References to self in the threads will prevent this object from ever getting garbage collected.
- This would only put one STOP_JOB. We would need one per thread if this is how we decide to cleanup.

Reply:
Oh god, yes. I'll fix this.

Reply:
Thought about this, but the worker threads will spend their time blocking on `queue.get()` and will not be able to receive the `threading.Event` signal until they've received the next job on the queue. I've modified the code to push the number of STOP jobs corresponding to the number of threads that are still alive at the time `stop()` is called.

Reply:
Could also give the get a timeout, and if no job gotten looped around, checking the event

Reply:
For example:
```py
job = self._queue.get(timeout=0.5)
if not job:
    continue
if stop_event.is_set()
```

Reply:
Now that the code is putting the right number of stop jobs on the queue everything seems to be working as expected. I can `start()` and `stop()` the service multiple times and there seems to be no thread leakage.

Reply:
I'm resolving this conversation, which is a duplicate of https://github.com/invoke-ai/InvokeAI/pull/5225#pullrequestreview-1776759508 .